File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        
        if list1 is None or list2 is None:
            return list1 if list2 is None else list2
        
        head = None
        curr = None
        left = list1
        right = list2
        
        if right.val >= left.val:
            head = left
            curr = head
            left = left.next
        else:
            head = right
            curr = head
            right = right.next

        while left is not None and right is not None:
            if right.val >= left.val:
                curr.next = left
                left = left.next
            else:
                curr.next = right
                right = right.next
            curr = curr.next



        if left is not None:
            curr.next = left
        elif right is not None:
            curr.next = right
        
        return head

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func removeNthFromEnd(_ head: ListNode?, _ n: Int) -> ListNode? {
        if head == nil{return nil}
        let tail = reverseNode(head!)
        var l: ListNode? = nil
        var r = tail

        var c = 1
        while c < n{
            l = r
            r = r?.next
            c += 1
        }

        if l != nil{
            l?.next = r?.next
        }
        
        return reverseNode(l == nil ? r!.next : tail!)

    }

    func reverseNode(_ node: ListNode?) -> ListNode?{
        var tail: ListNode? = nil
        var curr = node
        while curr != nil{
            var next = curr?.next
            curr?.next = tail
            tail = curr
            curr = next
        }
        return tail
    }
}